Federated Ingress Controller

quinton-hoole opened the pull request:
Based on new federated controller libraries. 

cc @kubernetes/sig-cluster-federation @mfanjie @nikhiljindal @mwielgus @mml @madhusudancs FYI

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30419)

<!-- Reviewable:end -->

Commit:
Add Federated Ingress Controller.

mfanjie commented on federation/pkg/federation-controller/ingress/ingress-controller.go:
```diff
@@ -0,0 +1,301 @@
+	ic.ingressInformerStore, ic.ingressInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").List(options)
```
should use `api.NamespaceAll`?

quinton-hoole replied:
Yes, agreed.  I couldn't remember the correct incantation, so I added that placeholder instead :-)

mfanjie commented on federation/pkg/federation-controller/ingress/ingress-controller.go:
```diff
@@ -0,0 +1,301 @@
+	ingressFederatedInformer util.FederatedInformer
+	// For updating members of federation.
+	federatedUpdater util.FederatedUpdater
+	// Definitions of ingresss that should be federated.
```
ingress->ingresses

quinton-hoole replied:
Agreed, done.

mfanjie commented on federation/pkg/federation-controller/ingress/ingress-controller.go:
```diff
@@ -0,0 +1,301 @@
+
+func (ic *IngressController) deliverIngressObj(obj interface{}, delay time.Duration, trial int64) {
+	ingress := obj.(*extensions.Ingress)
+	ic.deliverIngress(ingress.Name, delay, trial)
```
should we put namespace+name as the key?

quinton-hoole replied:
Agreed.  Will do.   The controller that this is modelled on is for namespaces themselves, which are not scoped by namespace, hence this.

mfanjie commented on federation/pkg/federation-controller/ingress/ingress-controller.go:
```diff
@@ -0,0 +1,301 @@
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, 0, 0) }))
+
+	// Federated informer on ingresss in members of federation.
+	ic.ingressFederatedInformer = util.NewFederatedInformer(
```
The user agent name of federatedInformer is `userAgentName     = "federation-service-controller"`, we need change this as a more comment name.

quinton-hoole replied:
Agreed, but fix in a separate PR, to keep this PR clean.

jianhuiz commented on federation/pkg/federation-controller/ingress/ingress-controller.go:
```diff
@@ -0,0 +1,301 @@
+		},
+		&extensions.Ingress{},
+		controller.NoResyncPeriodFunc(),
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, 0, 0) }))
```
Do we still need to handle `DeletedFinalStateUnknown` case?

quinton-hoole commented:
This is ready for review now.  Unit tests pass, although I'm sure we'll find some issues when tested against the new e2e tests in https://github.com/kubernetes/kubernetes/pull/29773.

I will add the following in small followup PR's:

1. DNS record creation
2. namespace handling
3. sharing a public IP across shards

cc @mml @nikhiljindal @madhusudancs

quinton-hoole commented:
Whoops!  It seems I messed up a branch merge.  Until I fix that up, reviewers need only look at 

1.  federation/pkg/federation-controller/ingress
2.  federation/cmd/federation-controller-manager/app/controllermanager.go

quinton-hoole commented:
OK, I fixed my merge mess-up.

mwielgus commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -0,0 +1,325 @@
+// error (like failure to create).
+type ingressItem struct {
+	ingress      string
+	prevAttempts int64
```
you probably don't need prev attempts (and this struct at all).

quinton-hoole replied:
Done.

mwielgus commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -0,0 +1,325 @@
+	}
+}
+
+func backoff(trial int64) time.Duration {
```
backoff is now handled with util/lifecycle/backoff. please take a look at namespace controller.

quinton-hoole replied:
Done.

mwielgus commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -0,0 +1,325 @@
+func (ic *IngressController) reconcileIngress(ingress string) {
+	glog.V(4).Infof("Reconciling ingress %q", ingress)
+	if !ic.isSynced() {
+		ic.deliverIngress(ingress, ic.clusterAvailableDelay, false)
```
return if not synced

quinton-hoole replied:
Done.

quinton-hoole commented:
Thanks for the review @mwielgus .  I've:

1. addressed all of your comments
2. added support for namespaces (as we discussed)
3. fixed a bug in the create and update logic around ObjectMeta (as discussed)
4. added logic to share a single global IP across all shards as described [here](https://gist.github.com/bprashanth/52648b2a0b6a5b637f843e7efb2abc97)

PTAL.  I think this is ready for LGTM now. All unit tests, and manual e2e tests passed. Automated e2e tests are lined up in https://github.com/kubernetes/kubernetes/pull/29773 which should merge in the next couple of days.

quinton-hoole commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 62 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-62:-KPko-I-DbavhUc1-KtZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L62)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed, done.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 96 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-96:-KPkoSTyq4i153DTX7-l:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L96)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Yes, agreed.  I couldn't remember the correct incantation, so I added that placeholder instead :-)
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 107 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-107:-KPkoWWQg2eD8mG6OL69:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L107)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed, but fix in a separate PR, to keep this PR clean.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 176 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-176:-KPkoZyWzAWFSEsslhIx:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L176)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed.  Will do.   The controller that this is modelled on is for namespaces themselves, which are not scoped by namespace, hence this.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 80 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KPUY_wXsqf5H0npP5Hy-r5-80:-KPkoJjk9jEK7QQVqpU0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e65ff45a3f89020739036f77e9fe9e0fc0f1af89/federation/pkg/federation-controller/ingress/ingress_controller.go#L80)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Done.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 255 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KPUY_wXsqf5H0npP5Hy-r5-255:-KPkol39L5g90aCggRsn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e65ff45a3f89020739036f77e9fe9e0fc0f1af89/federation/pkg/federation-controller/ingress/ingress_controller.go#L255)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419)*
<!-- Sent from Reviewable.io -->

mwielgus commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -0,0 +1,360 @@
+	"k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	// api_v1 "k8s.io/kubernetes/pkg/api/v1"
```
remove

quinton-hoole replied:
Yup, done.

mwielgus commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -0,0 +1,360 @@
+				// the equivalence comparison to cut down on unnecessary updates.
+				glog.V(4).Infof("Ingress %s in cluster %s needs an update: cluster ingress %v is not equivalent to federated ingress %v", ingress, cluster.Name, clusterIngress, desiredIngress)
+				// We need to use server-created fields from the cluster, not the desired object when updating.
+				desiredIngress.ObjectMeta.ResourceVersion = clusterIngress.ObjectMeta.ResourceVersion
```
Do we want to copy this?

quinton-hoole replied:
Good catch - I don't know what I was thinking there.  Will fix.

quinton-hoole replied:
Oh no wait, I was correct.  When updating (the cluster ingress), one needs to provide the resource version and UID.

mwielgus replied:
Why? In namespace controller i don't do that and it works. Also if you update an object using kubectl you don't need any of these fields.

quinton-hoole replied:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md

The resourceVersion is changed by the server every time an object is modified. If resourceVersion is included with the PUT operation the system will verify that there have not been other successful mutations to the resource during a read/modify/write cycle, by verifying that the current value of resourceVersion matches the specified value.

mwielgus replied:
The thing here is that we don't care about the previous content of the object here. We could overwrite the content every time we reconciliate an object but we do this only when something changed for performance reason.

quinton-hoole replied:
Yes, you're right, I had originally assumed that resourceVersion was mandatory for PUT's, but it's not.  I'll remove this in a followup PR.

mwielgus commented:
minor comments, LGTM assuming a clean-up PR is submitted shortly.

quinton-hoole commented:
Fixed the nit, squashed, and checked out the e2e test failure, which looks like a flake - the test cluster never came up.  Applying LGTM as per review feedback above.

quinton-hoole commented:
Rebased.  Re-applying LGTM.
